src/logics/ArithLogic.cc

```diff
@@ -194,6 +194,30 @@ bool ArithLogic::isLinearTerm(PTRef tr) const {
     return false;
 }
 
+bool ArithLogic::isNonlin(PTRef tr) const {
```

Comment:
Shouldn't this answer the question only for the current term?
If we need recursion into subterms, it should be implemented using a visitor.

Reply:
Initially I wanted it to be this way, but unfortunately in LASolver.cc we register bigger terms first.

So instead of registering `(/ x y)` we register `(* -1 (/ x y))` which LASolver assertions have problem with

Reply:
I'm not sure what you mean by using a `visitor`
I looked through the TermVisitor class, but it seems rather complicated, it means that we would need to implement the whole child class to check if LRA/LIA formula has nonlinearity which seems a bit as an overkill
Or did I got the idea wrong a bit?

Reply:
The visitor are exactly for that, to collect information of you need to traverse the whole term.

But I think we should not need to traverse whole terms to detect if somewhere inside is a nonlinear operation.

There should be one place, in LASolver, where you know you are dealing with a polynomial, so there you can check if it has degree more than one.

Reply:
> Initially I wanted it to be this way, but unfortunately in LASolver.cc we register bigger terms first.
> 
> So instead of registering `(/ x y)` we register `(* -1 (/ x y))` which LASolver assertions have problem with
> 
> And I think if LASolver already starts working it should not accept any Nonlin terms for the reason of separating LA and NLA

What do you mean?
I think in `LASolver::initSolver` before calling `registerArithmeticTerm` you can check if the polynomial has any nonlinear monomials.

Reply:
Fixed